### DISCUSSION TITLE
When using tailwind which grid / layout should be used

### DISCUSSION THREAD

**Discussion by TomFreudenberg** [zauberzeug/nicegui#2555] [Q&A] [answered]
### Question

Hi, I work a lot with tailwindcss when using nodejs frameworks like nextjs.

Very nice to use this in Python (backends) as well now, so thanks to this library!

I work out an event driven starter at https://github.com/tokeo/tokeo and have integrated a controller to launch a web service driven by nicegui to render some stats or backend pages.

Therefor I am unsure if I "break" the design of nicegui or quasar when using mostly the `ui.element` and style them by tailwind.

This works pretty good - but - could someone please guide me if this also the "designed way"?

```python
    ui.query('.nicegui-content').style('display: block')

    with ui.element('div').tailwind('bg-white', 'py-24', 'sm:py-32').element:
        with ui.element('div').tailwind('mx-auto', 'max-w-7xl', 'px-6', 'lg:px-8').element:
            with ui.element('dl').tailwind('grid', 'grid-cols-1', 'gap-x-8', 'gap-y-16', 'text-center', 'lg:grid-cols-3').element:
                with ui.element('div').tailwind('mx-auto', 'flex', 'max-w-xs', 'flex-col', 'gap-y-4').element:
                    with ui.element('dt').tailwind('text-base', 'leading-7', 'text-gray-600').element:
                        ui.label('Minis Ready')
                    with ui.element('dd').tailwind('order-first', 'text-3xl', 'font-semibold', 'tracking-tight', 'text-gray-900', 'sm:text-5xl').element:
                        ui.label(str(sql_result.total_items))
                with ui.element('div').tailwind('mx-auto', 'flex', 'max-w-xs', 'flex-col', 'gap-y-4').element:
                    with ui.element('dt').tailwind('text-base', 'leading-7', 'text-gray-600').element:
                        ui.label('Assets under holding')
                    with ui.element('dd').tailwind('order-first', 'text-3xl', 'font-semibold', 'tracking-tight', 'text-gray-900', 'sm:text-5xl').element:
                        ui.label('$119 trillion')
                with ui.element('div').tailwind('mx-auto', 'flex', 'max-w-xs', 'flex-col', 'gap-y-4').element:
                    with ui.element('dt').tailwind('text-base', 'leading-7', 'text-gray-600').element:
                        ui.label('New users annually')
                    with ui.element('dd').tailwind('order-first', 'text-3xl', 'font-semibold', 'tracking-tight', 'text-gray-900', 'sm:text-5xl').element:
                        ui.label('46,000')
```

Results in:

![image](https://github.com/zauberzeug/nicegui/assets/410087/caef8dc2-3553-4c52-b446-86c3a7f2f43f)

and that is like expected.

---

**Question:**

In the above case I do not use the quasar elements like `ui.row()` and `ui.col()` nor `ui.grid()`. Is this OK or do I something faulty by design.

Thanks for your feedback
Tom

**Comment by falkoschindler** [comment, marked as answer]
I think it's totally fine to use the generic `ui.element`. NiceGUI's layout elements like `ui.row` and `ui.column` are basically flexbox wrappers and more targeted to non-web-developers. But if you're more familiar with `dl`, `dt` and `dd`, feel free to use them.

Personally, I find code like this easier to ready though:
```py
with ui.grid(rows=2).classes('grid-flow-col mx-auto py-24 gap-x-24 text-center'):
    ui.label('317').classes('text-5xl font-semibold tracking-tight text-gray-900')
    ui.label('Minis Ready').classes('text-base leading-7 text-gray-600')
    ui.label('$119 trillion').classes('text-5xl font-semibold tracking-tight text-gray-900')
    ui.label('Assets under holding').classes('text-base leading-7 text-gray-600')
    ui.label('46,000').classes('text-5xl font-semibold tracking-tight text-gray-900')
    ui.label('New users annually').classes('text-base leading-7 text-gray-600')
```
<img width="788" alt="Screenshot 2024-02-16 at 12 28 16" src="https://github.com/zauberzeug/nicegui/assets/5767091/4a067b1e-e3ff-4054-8dd9-5d6173e8d46e">

**Reply by TomFreudenberg** [reply]
Hey @falkoschindler, thanks for your Feedback and the optional code.

As written in the other thread, I created a helper class as well to shorten the code.

But yes, your code example looks well, too

One question left then: you use `.classes` instead of `.tailwind` but write tailwind classes in. Does that mean it isn't necessary anymore to use the `.tailwind` method or why are they exist both?

Thanks for clarifying
Tom

**Reply by falkoschindler** [reply]
@TomFreudenberg The `.tailwind` method was introduced to make Tailwind more discoverable and Pythonic to use. But simply using Tailwind classes via `.classes()` has always been possible and is actually the way we personally like to use Tailwind.

**Reply by TomFreudenberg** [reply]
@falkoschindler so I do and will introduce my users as well.

P.S.: In addition it would be great to get your feedback on https://github.com/zauberzeug/nicegui/discussions/1277#discussioncomment-8474631 if you have the time.
